sdk/search/Azure.Search.Documents/src/Batching/Publisher{T}.cs

```diff
@@ -0,0 +1,493 @@
+        /// </param>
+        /// <param name="batchActionSize">
+        /// The number of actions to group into a batch.
+        /// </param>
+        /// <param name="batchPayloadSize">
+        /// The number of bytes to use when tuning the behavior of the
+        /// publisher.
+        /// </param>
+        /// <param name="retryCount">
+        /// The number of times to retry a failed document.
+        /// </param>
+        /// <param name="publisherCancellationToken">
+        /// A <see cref="CancellationToken"/> to use when publishing.
+        /// </param>
+        public Publisher(
+            bool autoFlush,
+            TimeSpan? autoFlushInterval,
+            int? batchActionSize,
+            int? batchPayloadSize,
+            int? retryCount,
+            CancellationToken publisherCancellationToken)
+        {
+            AutoFlush = autoFlush;
+            AutoFlushInterval = autoFlushInterval;
+            PublisherCancellationToken = publisherCancellationToken;
+            BatchActionSize = batchActionSize ?? SearchIndexingBufferedSenderOptions<T>.DefaultBatchActionSize;
+            BatchPayloadSize = batchPayloadSize ?? SearchIndexingBufferedSenderOptions<T>.DefaultBatchPayloadSize;
+            RetryCount = retryCount ??  SearchIndexingBufferedSenderOptions<T>.DefaultRetryCount;
+
+            // Start the message loop
+            _readerLoop = Task.Run(ProcessMessagesAsync, publisherCancellationToken);
+        }
+
+        #region Message Producers - only called externally and do not touch state
+        /// <summary>
+        /// Add documents to be published.  This should only be called outside
+        /// of the Publisher itself.
+        /// </summary>
+        /// <param name="documents">The documents to publish.</param>
+        /// <param name="async">Whether to call sync or async.</param>
+        /// <param name="cancellationToken">Cancellation token.</param>
+        /// <returns>Task for adding the documents.</returns>
+        public async Task AddDocumentsAsync(IEnumerable<T> documents, bool async, CancellationToken cancellationToken = default)
+        {
+            EnsureNotDisposed();
+            await SendMessageInternal(Message.Publish(documents), async, cancellationToken).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Flush any pending documents.  This should only be called outside of
+        /// the Publisher itself.
+        /// </summary>
+        /// <param name="async">Whether to call sync or async.</param>
+        /// <param name="cancellationToken">Cancellation token.</param>
+        /// <returns>Task for adding the documents.</returns>
+        public async Task FlushAsync(bool async, CancellationToken cancellationToken = default)
+        {
+            await SendMessageInternal(Message.Flush(), async, cancellationToken).ConfigureAwait(false);
+            await _flushBarrier.WaitAsync(async, cancellationToken).ConfigureAwait(false);
```

Comment:
What if other thread calls _flushBarrier.Release faster than we enter the semaphore inside the WaitAsync?

Reply:
Good catch.  I'll make sure only one thread can own whether folks are allowed to await/release at a time.